### PR TITLE
ExecutionContext[EC] phantom for control, Future[EC, A]

### PR DESCRIPTION
**Pull request by S11001001**
First version of a variant of `scala.concurrent.Future`, extremely compatible with it, using phantom type parameters to deal with the `ExecutionContext` overload problem. [Described more fully in the `package` Scaladoc](https://github.com/digital-asset/daml/pull/7347/files#diff-3c8bd3ba0f6495c6afb1621ae212e71dR11-R63), with examples showing idioms and bad expressions rejected by type-checking [in the test files](https://github.com/digital-asset/daml/pull/7347/files#diff-be8b7b63c6d574ad52ec5fe805f01ced).

Some things like whether `ExecutionContextExecutor` needs to be fully supported and whether `Resource` should be tagged are not addressed in this PR; we should address it if people like the library enough to start using these Futures.

Inspired by #7205, with feedback from experiments in anger by @SamirTalwar-DA.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/ExecutionContextOf.scala
Comment (S11001001) [inline libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/ExecutionContextOf.scala:10]: We can sensibly incorporate `ExecutionContextExecutor` in this scheme by writing

```scala
  type T[+P, +A] <: A
```

and defining aliases accordingly. It would then be a little closer to `scalaz.Tag`, albeit with subtyping, as well.  I'm not sure support for such subtypes is needed, though.

File: libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala
Comment (S11001001) [inline libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala:22]: After writing the initial implementation and looking at it for a bit, I started to worry about the placement of `EC`.  I mean to place limits on the `map`, `flatMap`, &c operations called on this Future.  But the `EC` on a Future has nothing to do with the execution context in which *that future was evaluated*, it constrains only *the combinators on that Future, i.e. other Futures it may produce*.

My worry was "this EC on the Future has nothing to do with the Future; it's not really part of it".  I thought, "should I complicate the API by adding a kind of 'projection', a method you call before map, flatmap, &c to declare an EC for the following call?"  But then I remembered

> None of this *means* anything, none of it *is* anything.  It's all uninterpreted symbols.  This, in particular, isn't even structural, never mind operational; it's only nominal, so it's even more "meaning"-less than all programs already are.

Speaking *practically*, and abandoning a quest for "meaning" in phantoms, the only thing that matters is **if this purely nominal abstraction well-types the kinds of things we want to be accepted, and ill-types the kinds of things we want to reject** (without further annotation that you know what you're doing).  Any constraints and constraint rules that make that happen are worthy of consideration.

File: libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala
Comment (S11001001) [inline libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala:52]: I'm rethinking this; @SamirTalwar-DA do you think it should be globally visible (as it is now), or should you have to import a `Conversions._` type thing, or use an explicit `asScalaUtils` method (like `JavaConverters`)?

Comment (SamirTalwar-DA) [reply libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala:52]: I think it's fine. The odds of it showing up accidentally are near-zero as far as I can see.

Comment (leo-da) [reply libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala:52]: My personal preference would be explicit `asScalaUtils` call. We all know that implicit `scala.collection.JavaConversions` are deprecated in favor of more explicit `scala.collection.JavaConverters`, right?

Comment (S11001001) [reply libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala:53]: I'm going to go with the current global availability for first iteration. We can scope it down if it seems to be a problem.

**Review by SamirTalwar-DA — COMMENTED** [review]
I think that broadly, I'm happy with the changes. I assume you're planning on actually using this somewhere.

My only concern is with the packaging. I'm not a big fan of "utils" grab-bags; they tend to just accumulate things with no relation to each other. I'd prefer if this was its own Bazel package (e.g. _libs-scala/futures_ or _libs-scala/concurrent_), and the root Java package was renamed accordingly (e.g. `com.daml.futures` or `com.daml.concurrent`).

**Comment by S11001001** [comment]
> I'd prefer if this was its own Bazel package (e.g. _libs-scala/futures_ or _libs-scala/concurrent_)

I'm just watchful of the `ExecutionContextOf`, which will instantly become agnostic if we [do this](#discussion_r485255242).

But anyway, sure, @leo-da could you pick a name?

**Comment by SamirTalwar-DA** [comment]
What do you mean by "agnostic"?

**Comment by S11001001** [comment]
> I assume you're planning on actually using this somewhere.

@SamirTalwar-DA Since you went through the trouble of actually trying to use it instead of scribbling down type ideas, I have a better idea 😈 

> What do you mean by "agnostic"?

It won't really have anything to do with ExecutionContexts or the concurrent library at that point. It'd be the "subtype newtype with covariant tags" thing.

**Comment by SamirTalwar-DA** [comment]
Makes sense. At which point I guess you'll probably have `SubtypeNewtypeWithCovariantTags` in its own library, and `type ExecutionContext = …` in this one.

What I'm saying is moar tiny libraries plz. 🙃

**Comment by S11001001** [comment]
`concurrent` it is. 1ace4b4e7b6739bd56e08ce1b9aa1fd5b6033baa

**Comment by leo-da** [comment]
> > I'd prefer if this was its own Bazel package (e.g. _libs-scala/futures_ or _libs-scala/concurrent_)
> 
> I'm just watchful of the `ExecutionContextOf`, which will instantly become agnostic if we [do this](#discussion_r485255242).
> 
> But anyway, sure, @leo-da could you pick a name?

`TypeEnrichedExecutionContext`?

**Comment by S11001001** [comment]
@leo-da No, I meant the library; anyway it's called 'concurrent' now.

File: libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala
Comment (leo-da) [inline libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala:35]: If anyone wants `Future[Cat, Int]` from `sc.Future[Int]`, they should convert it explicitly.
What I mean is ```com.daml.concurrent.FutureOf#`future is any` ``` should not exist. I would prefer an explicit converter in the `com.daml.concurrent.FutureOf.Ops`.

**Review by leo-da — COMMENTED** [review]
I like it, except the implicit `Future` conversion... see my comment.

In any case, I would leave it to other reviewers to approve if they are fine with it.

File: libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala
Comment (S11001001) [reply libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala:35]: Already mentioned this in the previous thread, but https://github.com/digital-asset/daml/pull/7347#discussion_r489001143 for this PR.

But since you mentioned it again, I thought I would write down the case for not bothering with an explicit converter:

The analogy with Scala vs Java collections is faulty in important ways. That is because there are no operational differences between dc.Futures (the ones here) and sc.Futures, which is not true of, say, java.util.List vs scala.collection.mutable.Buffer (or any other choice you may have). You can view the subtyping hierarchy of ExecutionContext as follows:

```
            sc.ExecutionContext
                     |
                     |
          dc.ExecutionContext[Any]
                /           \
               /             \
dc.ExecutionContext[<complete *-kinded type lattice>]
               \             /
                \           /
        dc.ExecutionContext[Nothing]
```

Comparably, *structurally*, the only difference between an sc.Future and dc.Future is which subset of the above types you accept as a context argument:

```
          dc.Future[Nothing]
            /           \
           /             \
dc.Future[<inverted *-kinded type lattice>]
           \             /
            \           /
            dc.Future[Any]
                  |
                  |
              sc.Future
```

(more specific types are lower, wider types are higher, for both graphs)

With the top of the Future hierarchy accepting only the bottom of the ExecutionContext hierarchy, each more specific Future accepting a broader, less specific set of ExecutionContexts, until you reach dc.Future[Any] which accepts any dc.ExecutionContext, and then finally the bottom of the Future hierarchy, which is the only one that accepts the top of the ExecutionContext hierarchy (and indeed all other ExecutionContexts).

Explicit conversion barriers between Scala collections and Java collections, or Scala futures and Java futures for that matter, are desirable because they cannot have a relationship meeting the Liskov Substitution Principle.  They have different dynamic behavior, so neither can be a substitute for the other, i.e. there exist functions on each that simply do not work on the other.

By contrast, _all_ functions that operate on a dc.Future[T] argument are also functions that operate on a sc.Future argument in the exact same way; LSP holds, there are no dynamic behavior differences at all, and the only difference is the type of ExecutionContext it accepts (which is a contravariant position).

Honestly I tried to actually reify this type relationship directly in Scala, but the specifics were too tricky to immediately resolve. It is worth further noting, however, that whereas `ScalaConversions` require fancy wrappers (which may have bugs) with special runtime characteristics of their own, the conversion here literally does nothing. That is what you'd expect, given that there is nothing about reifying the type relationship that would cause any particular problems for implementation.

It being that an implicit conversion that does nothing is significantly less likely to cause trouble than one that does something, I'd like to see whether it actually causes any problems in practice before assuming it does and adding burdens to anyone who might want to try porting sc.Future code to interact with dc.Futures.

File: release/artifacts.yaml
Comment (SamirTalwar-DA) [inline release/artifacts.yaml:157]: Sort plz.

**Review by SamirTalwar-DA — APPROVED** [review]

**Comment by cocreature** [comment]
Nice work, thank you!

File: libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala
Comment (leo-da) [reply libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala:35]: Correct me if I am wrong. The goal of this entire exercise was to make choosing `ExecutionContext` more explicit. The implicit conversion makes it possible to pass `dc.Future[Cat, Int]` into a method that expects `sc.Future[Int]`. And this whole type safety goes away unnoticed by the compiler.

Comment (leo-da) [reply libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala:35]: it is not exactly like this, it is opposite you can stick an 'untyped' `sc.Future[Int]` into a method that expects `dc.Future[Cat, Int]` and the compiler will not notice this and execute this `sc.Future[Int]` in the `dc.ExecutionContext[Cat]` which might be not describable.

Comment (S11001001) [reply libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala:35]: `sc.Future[Int]`'s type says it can be run by any `ExecutionContext`. `dc.ExecutionContext[Cat]` is a member of that set.

Comment (S11001001) [reply libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala:35]: That is to say, you are correct in that you might not desire to execute that sc.Future against *that* ExecutionContext, despite its type saying that it is okay. But that is why I want to make it as easy as possible to get sc.Futures *into* the dc.Future hierarchy that is placed above it in the type hierarchy; a sc.Future is already a failure of typing in this way, so we might as well ease its transition into a more precisely typed form that *can* be sufficiently specific.

File: release/artifacts.yaml
Comment (S11001001) [reply release/artifacts.yaml:157]: #7433

File: libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala
Comment (SamirTalwar-DA) [reply libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala:35]: I'd be happy to make the conversions more explicit once we've migrated chunks of the code over and understand this a little more. Right now I want it to be easy to flip back and forth as necessary so I can just use it in one file.

**Pull request merged**
